src/System.Private.Reflection.Metadata.Ecma335/src/Permanent/ImmutableExtensions.cs

```diff
@@ -0,0 +1,294 @@
+                return true;
+            }
+
+            var collectionOfT = sequence as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                count = collectionOfT.Count;
+                return true;
+            }
+
+            var readOnlyCollection = sequence as IReadOnlyCollection<T>;
+            if (readOnlyCollection != null)
+            {
+                count = readOnlyCollection.Count;
+                return true;
+            }
+
+            count = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the number of elements in the specified sequence,
+        /// while guaranteeing that the sequence is only enumerated once
+        /// in total by this method and the caller.
+        /// </summary>
+        /// <typeparam name="T">The type of element in the collection.</typeparam>
+        /// <param name="sequence">The sequence.</param>
+        /// <returns>The number of elements in the sequence.</returns>
+        internal static int GetCount<T>(ref IEnumerable<T> sequence)
+        {
+            int count;
+            if (!sequence.TryGetCount(out count))
+            {
+                // We cannot predict the length of the sequence. We must walk the entire sequence
+                // to find the count. But avoid our caller also having to enumerate by capturing
+                // the enumeration in a snapshot and passing that back to the caller.
+                var list = new List<T>(sequence);
+                count = list.Count;
+                sequence = list;
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Tries to copy the elements in the sequence to the specified array,
+        /// if the sequence is a well-known collection type. Otherwise, does
+        /// nothing and returns <c>false</c>.
+        /// </summary>
+        /// <typeparam name="T">The type of element in the sequence.</typeparam>
+        /// <param name="sequence">The sequence to copy.</param>
+        /// <param name="array">The array to copy the elements to.</param>
+        /// <param name="arrayIndex">The index in the array to start copying.</param>
+        /// <returns><c>true</c> if the elements were successfully copied; <c>false</c> otherwise.</returns>
+        /// <remarks>
+        /// <para>
+        /// The reason we don't copy anything other than for well-known types is that a malicious interface
+        /// implementation of <see cref="ICollection{T}"/> could hold on to the array when its <see cref="ICollection{T}.CopyTo"/>
+        /// method is called. If the array it holds onto underlies an <see cref="ImmutableArray{T}"/>, it could violate
+        /// immutability by modifying the array.
+        /// </para>
+        /// </remarks>
+        internal static bool TryCopyTo<T>(this IEnumerable<T> sequence, T[] array, int arrayIndex)
+        {
+            // IList is the GCD of what the following 2 types implement.
```

Comment:
This one is a bit more interesting, in that I made a separate copy of the code. In fact, I think the idea of splitting these contents into a separate file is a bad one, and I should instead either break the ImmutableExtensions.cs file up into two files like I did with ImmutableArray_1.cs, or add some ifdefs.

@AArnott  Andrew, I'd like your opinion on this. Are you still the maintainer of this code?